src/devices/src/virtio/persist.rs

```diff
@@ -40,6 +45,25 @@ pub struct QueueState {
     next_used: Wrapping<u16>,
 }
 
+impl QueueState {
+    /// Does a sanity check against expected values.
+    pub fn sanity_check(&self, queue_max_size: u16) -> std::result::Result<(), Error> {
+        // Cannot use `q.is_valid()` because snapshot can happen at any time,
+        // including during device configuration/activation when fields are only
+        // partially configured.
+        // We can't even check if GuestAddresses are valid guest phys addresses because
+        // their configuration happens in two steps for the two u32 halves of the
+        // u64 address. This means a snapshot can capture them only partially configured.
```

Comment:
Only if the device status is active could we also check `q.is_valid()` though ?

Reply:
I looked into doing that, but the current design doesn't easily allow it.
The queues and their state are decoupled from the (virtio)MmioTransport where the `active` information is stored.

The queues are restored as part of device restoration, then after we have a device we proceed to re-register/restore MmioTransport state.

I don't see an elegant solution to doing it..